rpc/http.go

```diff
@@ -273,6 +273,14 @@ func newHTTPServerConn(r *http.Request, w http.ResponseWriter) ServerCodec {
 			return err
 		}
 		w.Header().Set("content-length", strconv.Itoa(len(encdata)))
+
+		// If this request is wrapped in a handler that might remove Content-Length (such
+		// as the automatic gzip we do in package node), we need to ensure the HTTP server
+		// doesn't perform chunked encoding. When timeout case, the chunked encoding might
+		// not be finished correctly, and some clients will refuse to process the response
+		// when the final chunk is missing.
+		w.Header().Set("transfer-encoding", "identity")
```

Comment:
Will this still use content-encoding gzip, but transfer-encoding identity?

Reply:
Yes content-encoding will be gzip. This solution fixes it for some clients, most notably the go http client (and python).

Curl and ethers.js fail with EoF. I initially thought it fixes ethers but seems gzip compression is disabled by default. After enabling it, it fails too.

```console
❯ curl -X POST -H "Content-Type: application/json" -H "Accept-Encoding: gzip" --data '{"method":"eth_blockNumber","id":1,"jsonrpc":"2.0"}' --output - -s http://127.0.0.1:8545 | gunzip -
gunzip: (stdin): unexpected end of file
```

```js
const { ethers } = require('ethers')
async function main() {
    const provider = new ethers.providers.JsonRpcProvider({ url: 'http://127.0.0.1:8545', allowGzip: true })
    const num = await provider.getBlockNumber()
    console.log(num)
}
main().then().catch((err) => console.log(err))
```
Result:
```console
❯ node test.js
node:zlib:465
      throw self[kError];
      ^

Error: unexpected end of file
    at Zlib.zlibOnError [as onerror] (node:zlib:189:17)
    at processChunkSync (node:zlib:455:12)
    at zlibBufferSync (node:zlib:178:12)
    at syncBufferWrapper (node:zlib:790:14)
    at IncomingMessage.<anonymous> (/Users/sina/dev/geth-scripts/node_modules/ethers/node_modules/@ethersproject/web/lib/geturl.js:78:81)
    at IncomingMessage.emit (node:events:539:35)
    at endReadableNT (node:internal/streams/readable:1345:12)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) {
  errno: -5,
  code: 'Z_BUF_ERROR'
}
```

Reply:
That said, I think this PR is strictly much better than status quo:
- If gzip compression is disabled it returns an error message on timeout across the board, both for messages and batch requests.
- Timing wise error is returned as soon as timeout hits and connection is terminated, instead of waiting until the method handler returns. This is also strictly an improvement because connection was only cut after the method handler returned before.
- When gzip is enabled the Go client will get the error message. Curl and ethers.js don't. But even there connection is cut sooner and there wasn't a response before anyway.

Reply:
Been thinking about it more, we could check for the special value `transfer-encoding` value `identity` in the gzip handler and skip compression. That would fix the JS problem.

Reply:
It works! both for JS and curl. Will push it now

Reply:
I'm not sure if there's still a benefit in keeping the length-tracking in gzipWriter. I think we can remove it?